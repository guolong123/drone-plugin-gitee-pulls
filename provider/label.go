package provider

import (
	"fmt"
	"net/url"

	"github.com/drone/go-scm/scm"
)

type LabelService interface {
	// List 仓库任务标签列表
	List(repo string) (*scm.Response, []*Label, error)
	// Get 获取指定任务标签
	Get(repo string, name string) (*scm.Response, *Label, error)
	// Create 创建任务标签
	Create(repo string, in Label) (*scm.Response, *Label, error)
	// Delete 删除任务标签
	Delete(repo string, name string) (*scm.Response, error)
	// Update 更新任务标签
	Update(repo string, name string, in Label) (*scm.Response, *Label, error)
}

type Label struct {
	Name  string `json:"name"`
	Color string `json:"color"`
}

type labelService struct {
	client *wrapper
}

func (l *labelService) List(repo string) (*scm.Response, []*Label, error) {
	path := fmt.Sprintf("repos/%s/labels", repo)
	var out []*Label
	res, err := l.client.do("GET", path, nil, &out)
	if err != nil {
		return res, nil, err
	}
	return res, out, nil
}

func (l *labelService) Get(repo string, name string) (*scm.Response, *Label, error) {
	path := fmt.Sprintf("repos/%s/labels/%s", repo, url.QueryEscape(name))
	var out *Label
	res, err := l.client.do("GET", path, nil, &out)
	if err != nil {
		return res, nil, err
	}
	return res, out, nil
}

func (l *labelService) Create(repo string, in Label) (*scm.Response, *Label, error) {
	path := fmt.Sprintf("repos/%s/labels", repo)
	var out *Label
	res, err := l.client.do("POST", path, in, &out)
	if err != nil {
		return res, nil, err
	}
	return res, out, nil
}

func (l *labelService) Delete(repo string, name string) (*scm.Response, error) {
	path := fmt.Sprintf("repos/%s/labels/%s", repo, name)
	res, err := l.client.do("DELETE", path, nil, nil)
	return res, err
}

func (l *labelService) Update(repo string, originalName string, in Label) (*scm.Response, *Label, error) {
	path := fmt.Sprintf("repos/%s/labels/%s", repo, originalName)
	var out *Label
	res, err := l.client.do("PUT", path, in, &out)
	if err != nil {
		return res, nil, err
	}
	return res, out, nil
}
