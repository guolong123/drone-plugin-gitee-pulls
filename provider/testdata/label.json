{
  "id": 135184972,
  "name": "drone-build/running",
  "color": "E6A23C",
  "repository_id": 19390416,
  "url": "https://gitee.com/api/v5/repos/kit101/demo1/labels/drone-build/running",
  "created_at": "2021-11-30T17:59:51+08:00",
  "updated_at": "2021-11-30T18:07:37+08:00"
}