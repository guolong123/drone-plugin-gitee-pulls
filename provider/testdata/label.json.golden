{
  "Name": "drone-build/running",
  "Color": "E6A23C"
}